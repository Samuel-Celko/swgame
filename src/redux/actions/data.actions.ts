import { IDataState } from "../../model";

export enum ActionTypes {
    DATA_GET_INIT = "[DATA] Get Init",
    DATA_GET_SUCCESS = "[DATA] Get Success",
    DATA_GET_ERROR = "[DATA] Get Error"
}

interface IDataGetInit {
    readonly type: ActionTypes.DATA_GET_INIT;
}

interface IDataGetSuccess {
    readonly type: ActionTypes.DATA_GET_SUCCESS;
    payload: IDataState;
}

interface IDataGetError {
    readonly type: ActionTypes.DATA_GET_ERROR;
    payload: any;
}

export type Actions = IDataGetInit | IDataGetSuccess | IDataGetError;

export function getDataInit(): IDataGetInit {
    return {
        type: ActionTypes.DATA_GET_INIT,
    };
}

export function getDataSuccess(response: IDataState): IDataGetSuccess {
    return {
        type: ActionTypes.DATA_GET_SUCCESS,
        payload: response,
    };
}

export function getDataError(response: any): IDataGetError {
    return {
        type: ActionTypes.DATA_GET_ERROR,
        payload: response,
    };
}