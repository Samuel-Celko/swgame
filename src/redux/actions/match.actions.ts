import { IMatchState } from "../../model";

export enum ActionTypes {
    MATCH_START_INIT = "[MATCH] Start Init",
    MATCH_START_SUCCESS = "[MATCH] Start Success",
    MATCH_START_ERROR = "[MATCH] Start Error",
}

interface IMatchStartInit {
    readonly type: ActionTypes.MATCH_START_INIT;
}

interface IMatchStartSuccess {
    readonly type: ActionTypes.MATCH_START_SUCCESS;
    payload: IMatchState;
}

interface IMatchStartError {
    readonly type: ActionTypes.MATCH_START_ERROR;
    payload: any;
}

export type Actions = IMatchStartInit | IMatchStartSuccess | IMatchStartError;

export function matchStartInit(): IMatchStartInit {
    return {
        type: ActionTypes.MATCH_START_INIT
    };
}

export function matchStartSuccess(response: IMatchState): IMatchStartSuccess {
    return {
        type: ActionTypes.MATCH_START_SUCCESS,
        payload: response,
    };
}

export function matchStartError(response: any): IMatchStartError {
    return {
        type: ActionTypes.MATCH_START_ERROR,
        payload: response,
    };
}