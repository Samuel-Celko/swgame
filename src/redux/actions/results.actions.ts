import { IMatchState } from "../../model";

export enum ActionTypes {
    RESULT_SAVE = "[RESULT] Save",
}

interface IResultSave {
    readonly type: ActionTypes.RESULT_SAVE;
    payload: IMatchState;
}

export type Actions = IResultSave;

export function resultSave(response: IMatchState): IResultSave {
    return {
        type: ActionTypes.RESULT_SAVE,
        payload: response,
    };
}
