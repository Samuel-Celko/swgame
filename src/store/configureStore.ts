import { createBrowserHistory } from "history";
import { applyMiddleware, compose, createStore } from "redux";
import createSagaMiddlware from "redux-saga";
import { routerMiddleware } from "connected-react-router";

import createRootReducer from "../redux/reducers";
import { watcherSaga } from "./sagas";

const history = createBrowserHistory();
const sagaMiddleware = createSagaMiddlware();

const getAppliedMiddleware = (history: any) => {
    if (process.env.NODE_ENV === "development") {
        return applyMiddleware(
            sagaMiddleware,
            routerMiddleware(history)
        );
    }
    return applyMiddleware(
        sagaMiddleware,
        routerMiddleware(history),
    )
}

const configureStore = (preloadedState?: any) => {
    const composeEnhancer = (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
    const store = createStore(
        createRootReducer(history),
        preloadedState,
        composeEnhancer(
            getAppliedMiddleware(history),
        )
    );
    sagaMiddleware.run(watcherSaga);

    return store;
}

export default configureStore;