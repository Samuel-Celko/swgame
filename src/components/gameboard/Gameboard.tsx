import React from "react";
import { push } from "connected-react-router";
import { useSelector, useDispatch } from "react-redux";

import { matchStartInit } from "../../redux/actions/match.actions";
import Card from "../card/Card";
import styles from "./Gameboard.module.scss";
import { IMatchState } from "../../model";
import { IStoreState } from "../../redux/reducers";

const baseClass = "gameboard";

const Gameboard: React.FC = () => {
  const dispatch: any = useDispatch();
  const matchData: IMatchState = useSelector(
    (state: IStoreState) => state.match
  );


  return (
    <div className={styles[baseClass]}>
      <div className={styles[`${baseClass}__play-area`]}>
        <div className={styles[`${baseClass}__play-area__red`]}>
          {matchData.ready && <Card data={matchData.red} />}
        </div>
        <div className={styles[`${baseClass}__play-area__blue`]}>
          {matchData.ready && <Card data={matchData.blue} />}
        </div>
      </div>
      <div className={styles[`${baseClass}__actions`]}>
        <button onClick={() => dispatch(matchStartInit())}>Begin Battle</button>
      </div>

      <div className={styles[`${baseClass}__actions`]}>
        <button onClick={() => dispatch(push("/results"))}>Show Results</button>
      </div>
    </div>
  );
};

export default Gameboard;
