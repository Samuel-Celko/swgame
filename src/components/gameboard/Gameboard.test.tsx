import React from "react";
import { Provider } from "react-redux";
import Adapter from "enzyme-adapter-react-16";
import configureStore from "redux-mock-store";
import { act } from "react-dom/test-utils";
import { mount, configure } from "enzyme";
import renderer from "react-test-renderer";

import Gameboard from "./Gameboard";
import { matchStartInit } from "../../redux/actions/match.actions";

const mockStore = configureStore([]);
configure({ adapter: new Adapter() });

describe("Gameboard", () => {
  let store: any;
  let component: React.FC | any;

  beforeEach(() => {
    store = mockStore({
      match: { ready: false }
    });

    store.dispatch = jest.fn();

    act(() => {
      component = mount(
        <Provider store={store}>
          <Gameboard />
        </Provider>
      );
    });
  });

  it("should render the component with a redux store", () => {
    const mockComponent = renderer
      .create(
        <Provider store={store}>
          <Gameboard />
        </Provider>
      )
      .toJSON();

    expect(mockComponent).toMatchSnapshot();
  });

  it("should render 2 card components if match is ready", () => {
    const store = mockStore({
      match: {
        ready: true,
        red: { data: {} },
        blue: { data: {} }
      }
    });

    const wrapper = mount(
      <Provider store={store}>
        <Gameboard />
      </Provider>
    );

    expect(wrapper.find("Card")).toHaveLength(2);
  });

  it("should not render any card components if match is not ready", () => {
    const store = mockStore({
      match: {
        ready: false
      }
    });

    const wrapper = mount(
      <Provider store={store}>
        <Gameboard />
      </Provider>
    );

    expect(wrapper.find("Card")).toHaveLength(0);
  });

  it("should dispatch matchStartInit action on button click", () => {
    const button = component.find("button");

    expect(button).toHaveLength(1);
    button.simulate("click");
    expect(store.dispatch).toHaveBeenCalledWith(matchStartInit());
  });
});
