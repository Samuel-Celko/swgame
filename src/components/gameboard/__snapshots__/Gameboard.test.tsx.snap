// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gameboard should render the component with a redux store 1`] = `
<div
  className="gameboard"
>
  <div
    className="gameboard__play-area"
  >
    <div
      className="gameboard__play-area__red"
    />
    <div
      className="gameboard__play-area__blue"
    />
  </div>
  <div
    className="gameboard__actions"
  >
    <button
      onClick={[Function]}
    >
      Begin Battle
    </button>
  </div>
</div>
`;
